File: examples/python/dispatch_model.py
"""Small Pyomo and HiGHS dispatch example for the COMPASS book."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Iterable

import pyomo.environ as pyo
from pyomo.contrib.appsi.base import TerminationCondition
from pyomo.contrib.appsi.solvers import Highs


@dataclass(frozen=True)
class Generator:
    name: str
    marginal_cost: float
    capacity: float

    def __post_init__(self) -> None:
        if self.marginal_cost < 0:
            raise ValueError("marginal_cost must be non-negative")
        if self.capacity < 0:
            raise ValueError("capacity must be non-negative")


def merit_order_dispatch(
    demand: float, generators: Iterable[Generator]
) -> list[tuple[str, float]]:
    if demand < 0:
        raise ValueError("demand must be non-negative")

    demand_value = float(demand)
    items = list(generators)

    if not items:
        if demand_value <= 1e-9:
            return []
        raise ValueError("demand exceeds available capacity")

    names = [generator.name for generator in items]
    if len(set(names)) != len(names):
        raise ValueError("generator names must be unique")

    model = pyo.ConcreteModel()
    model.generators = pyo.Set(initialize=range(len(items)))
    model.output = pyo.Var(
        model.generators,
        within=pyo.NonNegativeReals,
        bounds=lambda _, i: (0.0, items[i].capacity),
    )
    model.demand_balance = pyo.Constraint(
        expr=sum(model.output[i] for i in model.generators) == demand_value
    )
    model.total_cost = pyo.Objective(
        expr=sum(items[i].marginal_cost * model.output[i] for i in model.generators),
        sense=pyo.minimize,
    )

    solver = Highs()
    solver.config.load_solution = False
    results = solver.solve(model)

    if results.termination_condition in {
        TerminationCondition.infeasible,
        TerminationCondition.infeasibleOrUnbounded,
    }:
        raise ValueError("demand exceeds available capacity")
    if results.termination_condition != TerminationCondition.optimal:
        raise RuntimeError(
            "dispatch optimisation failed with status "
            f"{results.termination_condition}"
        )

    solver.load_vars()

    dispatch: list[tuple[str, float]] = []

    for i, generator in enumerate(items):
        output = pyo.value(model.output[i])
        if output > 1e-7:
            dispatch.append((generator.name, output))

    return dispatch


def total_cost(
    dispatch: Iterable[tuple[str, float]], generators: Iterable[Generator]
) -> float:
    costs = {generator.name: generator.marginal_cost for generator in generators}
    return sum(costs[name] * output for name, output in dispatch)


def demo() -> tuple[list[tuple[str, float]], float]:
    generators = [
        Generator("wind", 0.0, 35.0),
        Generator("solar", 3.0, 25.0),
        Generator("gas", 75.0, 60.0),
    ]
    dispatch = merit_order_dispatch(80.0, generators)
    return dispatch, total_cost(dispatch, generators)


if __name__ == "__main__":
    dispatch_result, dispatch_cost = demo()
    print(f"Dispatch: {dispatch_result}")
    print(f"Total cost: {dispatch_cost:.2f}")

File: examples/julia/dispatch_model.jl
module DispatchModel

using JuMP
import HiGHS

export Generator, merit_order_dispatch, total_cost

struct Generator
    name::String
    marginal_cost::Float64
    capacity::Float64

    function Generator(name::AbstractString, marginal_cost::Real, capacity::Real)
        marginal_cost < 0 && throw(ArgumentError("marginal_cost must be non-negative"))
        capacity < 0 && throw(ArgumentError("capacity must be non-negative"))
        new(String(name), Float64(marginal_cost), Float64(capacity))
    end
end

function merit_order_dispatch(demand::Real, generators)
    demand < 0 && throw(ArgumentError("demand must be non-negative"))

    demand_value = Float64(demand)
    items = collect(generators)

    if isempty(items)
        demand_value <= 1e-9 && return Pair{String, Float64}[]
        throw(ArgumentError("demand exceeds available capacity"))
    end

    names = [generator.name for generator in items]
    length(unique(names)) == length(names) ||
        throw(ArgumentError("generator names must be unique"))

    model = Model(HiGHS.Optimizer)
    set_silent(model)

    @variable(model, 0 <= output[i in eachindex(items)] <= items[i].capacity)
    @constraint(model, sum(output[i] for i in eachindex(items)) == demand_value)
    @objective(
        model,
        Min,
        sum(items[i].marginal_cost * output[i] for i in eachindex(items))
    )

    optimize!(model)

    status = termination_status(model)
    if status == JuMP.MOI.INFEASIBLE
        throw(ArgumentError("demand exceeds available capacity"))
    elseif status != JuMP.MOI.OPTIMAL
        throw(ErrorException("dispatch optimisation failed with status $status"))
    end

    dispatch = Pair{String, Float64}[]

    for (i, generator) in pairs(items)
        generation = value(output[i])
        if generation > 1e-7
            push!(dispatch, generator.name => generation)
        end
    end

    return dispatch
end

function total_cost(dispatch, generators)
    costs = Dict(generator.name => generator.marginal_cost for generator in generators)
    return sum(costs[item.first] * item.second for item in dispatch)
end

end

if abspath(PROGRAM_FILE) == @__FILE__
    using .DispatchModel

    generators = [
        Generator("wind", 0.0, 35.0),
        Generator("solar", 3.0, 25.0),
        Generator("gas", 75.0, 60.0),
    ]

    dispatch = merit_order_dispatch(80.0, generators)
    println("Dispatch: ", dispatch)
    println("Total cost: ", round(total_cost(dispatch, generators); digits = 2))
end
